src/main/kotlin/com/haroldadmin/cnradapter/ResponseParser.kt

```diff
@@ -0,0 +1,94 @@
+package com.haroldadmin.cnradapter
+
+import okhttp3.ResponseBody
+import retrofit2.Converter
+import retrofit2.HttpException
+import retrofit2.Response
+import java.io.IOException
+import java.lang.reflect.Type
+
+internal const val STATUS_NO_CONTENT = 204
+
+/**
+ * Maps a [Response] to a [NetworkResponse].
+ *
+ * A [Response] can either be successful or unsuccessful.
+ *
+ * For unsuccessful responses:
+ * - Try to parse the error body using [errorConverter].
+ * - If error body is parsed successfully, return it as [NetworkResponse.Error.ServerError]
+ * - Otherwise, assume we ran into an unknown error (probably related to serialization)
+ * and return [NetworkResponse.Error.UnknownError]
+ *
+ * For successful responses:
+ *
+ * If response body is null:
+ * - If [successType] is [Unit], return [NetworkResponse.Success] with [Unit] as the body
+ * - If [Response.code] is [STATUS_NO_CONTENT], return [NetworkResponse.Success] with [Unit] as the body
+ * - Else return a [NetworkResponse.Error.ServerError] with a null body
+ *
+ * If response body is not null:
+ * - Return [NetworkResponse.Success] with the parsed body
+ *
+ * @param errorConverter Retrofit provided body converter to parse the error body of the response
+ * @return A subtype of [NetworkResponse] based on the response of the network request
+ */
+@Suppress("UNCHECKED_CAST")
+internal fun <S, E> Response<S>.toNetworkResponse(
+    successType: Type,
+    errorConverter: Converter<ResponseBody, E>
+): NetworkResponse<S, E> {
+
+    if (!isSuccessful) {
```

Comment:
Can this be split into a success parsing function and an error parsing function?

Reply:
Sure, will make it easier to follow for any readers.